### DISCUSSION TITLE
First-class support for C++ as a scripting language -- what would it take?

### DISCUSSION THREAD

**Discussion by vittorioromeo** [godotengine/godot-proposals#9008] [Scripting]
A picture (mockup) is worth a thousand words. I want this:

![godotmockup](https://github.com/godotengine/godot-proposals/assets/918661/a6aab1d8-82ad-4de6-855e-be66ceb5819c)

When the script file is created, it opens up either in the internal text editor or in one I specify in Godot's settings, just like for C#. 

When the project needs to be built/executed, all C++ files are compiled into a hot-reloadable shared library using the system compiler (overridable in Godot's settings), and the library is loaded on application startup.

Basically, I want the same workflow that Godot offers for C#, but for C++ instead. **What would it take for this to happen?**

I can think of a few individual tasks:

- UI work to support the new language
    - Update "add script" widget
    - Update Godot settings to add C++ compiler selection
    - Update internal text editor to support C++
- API work
    - Expose a nice and clean C++ API to the end users, closely mimicking C# but also taking advantage of C++'s value semantics and Modern C++ features
- Internal work
    - Add ability to compile C++ files from the editor into a shared library
    - Add support to dynamically load C++ shader libraries on application build/run cycle
    
This seems very doable. Is there something I am missing?

**Comment by Calinou** [comment]
Thanks for the proposal! Consolidating in https://github.com/godotengine/godot-proposals/issues/3367, https://github.com/godotengine/godot-proposals/issues/573 and https://github.com/godotengine/godot-proposals/issues/119.

As you can see in the linked proposals, this is a very complex task that would require bespoke code to be written as C# doesn't go through GDExtension.
